### PR TITLE
Resume interrupted connections

### PR DESCRIPTION
**Pull request by tempusfrangit**
Resume interrupted connections. This makes the client.Client an interface as a test seam (so we can easily mock it). Additionally this now allows pget to resume downloads if a connection is interruped before the chunk is delivered. This should reduce the potential for unexpectedEOF errors being bubbled up to the end user.

### CODE REVIEW

File: pkg/download/consistent_hashing.go [hunk 1]
```diff
@@ -205,11 +208,6 @@ func (m *ConsistentHashingMode) downloadRemainingChunks(ctx context.Context, url
 					// for the specified chunk instead of the whole file.
 					if errors.Is(err, client.ErrStrategyFallback) {
 						// TODO(morgan): we should indicate the fallback strategy we're using in the logs
-						logger.Info().
-							Str("url", urlString).
-							Str("type", "chunk").
-							Err(err).
-							Msg("consistent hash fallback")
```
Comment (tempusfrangit) [inline pkg/download/consistent_hashing.go:212]:
```suggestion
						logger.Info().
							Str("url", urlString).
							Str("type", "chunk").
							Err(err).
							Msg("consistent hash fallback")
```
add this back in.

File: pkg/download/consistent_hashing.go [hunk 2]
```diff
@@ -210,6 +216,7 @@ func (m *ConsistentHashingMode) downloadRemainingChunks(ctx context.Context, url
 							Str("type", "chunk").
 							Err(err).
 							Msg("consistent hash fallback")
+
```
Comment (timfallmk) [inline pkg/download/consistent_hashing.go:219]: Extraneous?
```suggestion
```

File: pkg/download/consistent_hashing.go [hunk 3]
```diff
@@ -120,7 +120,7 @@ func (m *ConsistentHashingMode) Fetch(ctx context.Context, urlString string) (io
 			logger.Warn().
 				Int("connection_interrupted_at_byte", n).
 				Msg("Resuming Chunk Download")
-			_, err = resumeDownload(firstChunkResp.Request, buf[n:], m.Client, int64(n))
+			_, err = resumeDownload(firstChunkResp.Request, buf[n:contentLength], m.Client, int64(n))
```
Comment (tempusfrangit) [inline pkg/download/consistent_hashing.go:123]:
```suggestion
 			n, err = resumeDownload(firstChunkResp.Request, buf[n:], m.Client, int64(n))
```

File: pkg/download/consistent_hashing.go [hunk 4]
```diff
@@ -231,7 +231,7 @@ func (m *ConsistentHashingMode) downloadRemainingChunks(ctx context.Context, url
 					logger.Warn().
 						Int("connection_interrupted_at_byte", n).
 						Msg("Resuming Chunk Download")
-					_, err = resumeDownload(resp.Request, buf[n:], m.Client, int64(n))
+					_, err = resumeDownload(resp.Request, buf[n:contentLength], m.Client, int64(n))
```
Comment (tempusfrangit) [inline pkg/download/consistent_hashing.go:233]:
```suggestion
			n, err = resumeDownload(resp.Request, buf[n:contentLength], m.Client, int64(n))
```

**Pull request merged**